x/ibc-rate-limit/rate_limit.go

```diff
@@ -0,0 +1,96 @@
+package ibc_rate_limit
+
+import (
+	"encoding/json"
+	wasmkeeper "github.com/CosmWasm/wasmd/x/wasm/keeper"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	"github.com/cosmos/ibc-go/v3/modules/core/exported"
+	"github.com/osmosis-labs/osmosis/v11/x/ibc-rate-limit/types"
+)
+
+func CheckRateLimits(ctx sdk.Context, wasmKeeper *wasmkeeper.Keeper,
+	msgType, contract string,
+	channelValue sdk.Int, sourceChannel, denom string,
+	amount string,
+) error {
+	contractAddr, err := sdk.AccAddressFromBech32(contract)
+	if err != nil {
+		return err
+	}
+
+	sendPacketMsg, _ := BuildWasmExecMsg(
+		msgType,
+		sourceChannel,
+		denom,
+		channelValue,
+		amount,
+	)
+
+	contractKeeper := wasmkeeper.NewDefaultPermissionKeeper(wasmKeeper)
+	_, err = contractKeeper.Sudo(ctx, contractAddr, []byte(sendPacketMsg))
+
+	if err != nil {
+		return sdkerrors.Wrap(types.ErrRateLimitExceeded, err.Error())
+	}
+	return nil
+}
+
+type SendPacketMsg struct {
+	SendPacket RateLimitExecMsg `json:"send_packet"`
+}
+
+type RecvPacketMsg struct {
+	RecvPacket RateLimitExecMsg `json:"recv_packet"`
+}
+
+type RateLimitExecMsg struct {
+	ChannelId    string  `json:"channel_id"`
+	Denom        string  `json:"denom"`
+	ChannelValue sdk.Int `json:"channel_value"`
+	Funds        string  `json:"funds"`
+}
+
+func BuildWasmExecMsg(msgType, sourceChannel, denom string, channelValue sdk.Int, amount string) (string, error) {
+	content := RateLimitExecMsg{
+		ChannelId:    sourceChannel,
+		Denom:        denom,
+		ChannelValue: channelValue,
+		Funds:        amount,
+	}
+
+	var (
+		asJson []byte
+		err    error
+	)
+	switch {
+	case msgType == "send_packet":
+		msg := SendPacketMsg{SendPacket: content}
+		asJson, err = json.Marshal(msg)
```

Comment:
I don't think we can, cause then we'd have to define msg to be of a type that can be either SendPacket or RecvPacket.  Is there an easy way to do this in go?

Reply:
Makes sense!

A follow-up question: is there a reason for defining these messages as Go structs instead of generating them via proto files?

Reply:
These messages are defined in rust in the contract; the struct here is just a go representation of the message. For tests, we are probably going to be able to write those directly in rust and have the test use the chain as a packege. But I don't think there's a solution for dealing with this when calling the contract from the chain. 

An options is to try to auto generate the go types from the rust definitions using the json schema; I don't think this is necessary here since the messages are so simple (and so few), but we may want to do something like that for more complex contracts.

Reply:
Gotcha. Thanks